govcd/openapi.go

```diff
@@ -0,0 +1,481 @@
+	}
+
+	// rawJsonBodies contains a slice of all response objects and they must be formatted as a JSON slice (wrapped
+	// into `[]`, separated with semicolons) so that unmarshalling to specified `outType` works in one go
+	allResponses := `[` + strings.Join(rawJsonBodies, ",") + `]`
+
+	// Unmarshal all accumulated responses into `outType`
+	if err = json.Unmarshal([]byte(allResponses), &outType); err != nil {
+		return fmt.Errorf("error decoding values into type: %s", err)
+	}
+
+	return nil
+}
+
+// OpenApiGetItem is a low level OpenAPI client function to perform GET request for any item.
+// The urlRef must point to ID of exact item (e.g. '/1.0.0/edgeGateways/{EDGE_ID}')
+// It responds with HTTP 403: Forbidden - If the user is not authorized or the entity does not exist. When HTTP 403 is
+// returned this function returns "ErrorEntityNotFound: API_ERROR" so that one can use ContainsNotFound(err) to
+// differentiate when an objects was not found from any other error.
+func (client *Client) OpenApiGetItem(apiVersion string, urlRef *url.URL, params url.Values, outType interface{}) error {
+	util.Logger.Printf("[TRACE] Getting item from endpoint %s with expected response of type %s",
+		urlRef.String(), reflect.TypeOf(outType))
+
+	if !client.OpenApiIsSupported() {
+		return fmt.Errorf("OpenAPI is not supported on this VCD version")
+	}
+
+	req := client.newOpenApiRequest(apiVersion, params, http.MethodGet, urlRef, nil)
+	resp, err := client.Http.Do(req)
+	if err != nil {
+		return fmt.Errorf("error performing GET request to %s: %s", urlRef.String(), err)
+	}
+
+	// Bypassing the regular path using function checkRespWithErrType and returning parsed error directly
+	// HTTP 403: Forbidden - is returned if the user is not authorized or the entity does not exist.
+	if resp.StatusCode == http.StatusForbidden {
+		err := ParseErr(types.BodyTypeJSON, resp, &types.OpenApiError{})
+		resp.Body.Close()
+		return fmt.Errorf("%s: %s", ErrorEntityNotFound, err)
+	}
+
+	// resp is ignored below because it is the same as above
+	_, err = checkRespWithErrType(types.BodyTypeJSON, resp, err, &types.OpenApiError{})
+
+	// Any other error occurred
+	if err != nil {
+		return fmt.Errorf("error in HTTP GET request: %s", err)
+	}
+
+	if err = decodeBody(types.BodyTypeJSON, resp, outType); err != nil {
+		return fmt.Errorf("error decoding JSON response after GET: %s", err)
+	}
+
+	err = resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf("error closing response body: %s", err)
+	}
+
+	return nil
+}
+
+// OpenApiPostItem is a low level OpenAPI client function to perform POST request for any item.
+// The urlRef must point to POST endpoint (e.g. '/1.0.0/edgeGateways')
+// It handles synchronous and asynchronous tasks. When a task is synchronous - it will block until it is finished.
+func (client *Client) OpenApiPostItem(apiVersion string, urlRef *url.URL, params url.Values, payload, outType interface{}) error {
```

Comment:
Ok, can do. My point was that in any case one is able to get `outType` marshaled directly and does not have to know if that particular endpoint behaves synchronously or asynchronously. I did not find proper documentation how endpoints do that.

If we extract the async one (task) - one will have to lookup reference to new object from task.

Reply:
I have introduced additional `Sync` and `Async` versions of these functions. The main function without `Suffix` still handles both ways because it is much more convenient (one can get `outType` directly) without knowing if the endpoint in use works in "Sync" or "Async" way.